[gpu] strict driver and cuda version assignment

cjac opened the pull request:
Resolves Issues
* #1268 

gpu/install_gpu_driver.sh

* Driver version defaults to version in driver .run file if specified
* CUDA version defaults to version in cuda .run file if specified

* exclusively using .run file installation method for cuda and driver installation
* * Installing non-open driver from .run file on rocky8
* build nccl from source, since that is the only mechanism which supports all Dataproc OSs
* wrap expensive functions in completion checks to reduce re-run time when testing manually
* cache build results in GCS
* waiting on apt lock when it exists
* only install build dependencies if build is necessary
* fix problem with ops agent not installing ; using venv
* Installing gcc-12 on ubuntu22 to fix kernel driver FTBFS
* mark task completion by creating a file rather than setting a variable
* added functions to check and report secure-boot and os version details       
* if the correct metadata attributes are specified, pytorch will be installed, but not by default                      

gpu/manual-test-runner.sh
* order commands correctly
* point to origin rather than staging repo

gpu/test_gpu.py
* clearer test skipping logic

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
Tests took 43 minutes to run this time.  I think it's because the binary driver build cache was invalidated.  Let's see if this one takes more like 14 minutes.

cjac commented:
the intention is for this PR to resolve issue #1268

cjac commented on cloudbuild/Dockerfile:
```diff
@@ -9,16 +9,23 @@ COPY --chown=ia-tests:ia-tests . /init-actions
 # https://docs.bazel.build/versions/master/install-ubuntu.html
-ENV bazel_kr_path=/usr/share/keyrings/bazel-keyring.gpg
+ENV bazel_kr_path=/usr/share/keyrings/bazel-keyring.gpg \
+    bazel_version=7.4.0 \
```
Pinning to 7.4.0 is necessary since the release of and default set to version 8

cjac commented on cloudbuild/Dockerfile:
```diff
@@ -9,16 +9,23 @@ COPY --chown=ia-tests:ia-tests . /init-actions
-RUN /usr/bin/curl https://bazel.build/bazel-release.pub.gpg | \
-    gpg --dearmor -o "${bazel_kr_path}"
-RUN echo "deb [arch=amd64 signed-by=${bazel_kr_path}] http://storage.googleapis.com/bazel-apt stable jdk1.8" | \
+RUN /usr/bin/curl -s https://bazel.build/bazel-release.pub.gpg | \
```
reduce noise in docker build with the -s (silent) option

cjac commented on cloudbuild/Dockerfile:
```diff
@@ -9,16 +9,23 @@ COPY --chown=ia-tests:ia-tests . /init-actions
     apt-get clean
 
+# Set bazel-${bazel_version} as the default bazel alternative in this container
+RUN update-alternatives --install /usr/bin/bazel bazel /usr/bin/bazel-${bazel_version} 1 && \
```
The file /usr/bin/bazel does not exist in the bazel-${version} package ; set up an alternative and use it

cjac commented on cloudbuild/create-key-pair.sh:
```diff
@@ -0,0 +1,135 @@
+#!/bin/bash
```
this file copied from https://github.com/GoogleCloudDataproc/custom-images/tree/2024.11/examples/secure-boot

cjac commented on cloudbuild/presubmit.sh:
```diff
@@ -70,6 +70,7 @@ determine_tests_to_run() {
     changed_dir="${changed_dir%%/*}/"
     # Run all tests if common directories modified
     if [[ ${changed_dir} =~ ^(integration_tests|util|cloudbuild)/$ ]]; then
+      continue
```
must remove this before merging

cjac commented:
slightly longer than the 14 minutes I predicted, this run completed in 16:29

cjac commented on gpu/test_gpu.py:
```diff
@@ -299,7 +355,43 @@ def test_install_gpu_cuda_nvidia_with_spark_job(self, configuration, machine_suf
+    ("STANDARD", ["m", "w-0", "w-1"], GPU_T4, GPU_T4, "11.8", 'rocky', '2.0'),
+    ("STANDARD", ["m", "w-0", "w-1"], GPU_T4, GPU_T4, "12.4", 'rocky', '2.1'),
+    ("STANDARD", ["m", "w-0", "w-1"], GPU_T4, GPU_T4, "12.0", 'rocky', '2.2'),
+    ("KERBEROS", ["m", "w-0", "w-1"], GPU_T4, GPU_T4, "12.6", 'rocky', '2.2'),
```
We now exercise Kerberos in this test suite.  True, it's only tested on 2.2-rocky9 with CUDA 12.6, but we could change any of the STANDARD definitions above to KERBEROS to improve that set.

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
okay, taking a little break...

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

jayadeep-jayaraman commented:
There are too many changes in this single PR. Can the PR be broken into smaller logical chunks so that it can be reviewed better

cjac commented:
yeah, there has been a lot of change.  I'll re-order the functions a little.  This should reduce the delta significantly.  Testing my change before commit and push...

cjac commented:
/gcbrun

cjac commented:
blocked on #1283

cjac reviewed (commented):
These change are ready for review.

SurajAralihalli commented:
Referencing a related discussion: [https://github.com/GoogleCloudDataproc/initialization-actions/pull/1269#discussion_r1904787650](https://github.com/GoogleCloudDataproc/initialization-actions/pull/1269#discussion_r1904787650).

cjac commented:
> blocked on #1283

1283 has been merged ; continuing.

cjac commented:
> /usr/bin/nvidia-smi -i 0 --multi-instance-gpu 1

This command causes problems on 11.6.  I'm reviewing nvidia-smi -h for analog for that release.

cjac commented:
it wants an = for the last key/value pair but the first one must not have an =.  *shrug*

cjac commented:
/gcbrun

cjac commented:
rebased on master

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
okay.  I think this will do.  Kuldeep, over to you.

cjac commented:
I'm going to do a review of the squash and see if anything needs cleaning up.

cjac commented:
/gcbrun

cjac commented:
okay.  it's a lot, but it all looks okay and has been thoroughly and repeatedly tested. Nota Bene that 'continue' in presubmit.sh ; I'll remove that once I receive an LGTM.

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
Capacity Scheduler does not support gpu resources on Dataproc 2.0, so in order to use spark with yarn, fall back to Fair Scheduler ; I know there are concurrency concerns at high load, but that concern can be managed, and there is no other option on 2.0 clusters.

cjac commented:
/gcbrun

cjac commented:
* if cuda version and gpu driver version are not specified with metadata values, but the customer supplies a cuda runfile url, extract the driver and cuda version from the filename.
* There are inconsistencies between the driver version mentioned in the filename and the actual driver version that can be installed - I have mapped these out, so that if customer uses the latest minor version of any cuda release from 11.7 through 12.6.3 we have a known good driver for that cuda version without their having to think about it.  We also have likely to work configurations from 10.0 through 11.6 if the customer is willing to run an older dataproc subminor version with equally older kernels.
* Customer can explicitly specify driver version with either gpu-driver-version or gpu-driver-url metadata value
* selecting default cuda version based on dataproc image version
* The Yarn Capacity scheduler from hadoop [3.2.4](https://hadoop.apache.org/docs/r3.2.4/) does not recognize the gpu resource type and so must be replaced with the Fair scheduler.  NB: The Fair scheduler has problems at high concurrency when underprovisioned and will time out.
* NCCL is now built from source which takes a long first time for each cuda version per dataproc image the customer deploys. Which reminds me that
* tasks which used to take a long time to perform in previous versions of this script have had their results cached so that subsequent deployments of the same configuration will take significantly less time.

kuldeepkk-dev reviewed (commented):
LGTM. 

@cjac I would recommend getting one more review from the product team since there are lot of changes.

prince-cs commented on cloudbuild/presubmit.sh:
```diff
@@ -70,6 +70,7 @@ determine_tests_to_run() {
     changed_dir="${changed_dir%%/*}/"
     # Run all tests if common directories modified
     if [[ ${changed_dir} =~ ^(integration_tests|util|cloudbuild)/$ ]]; then
+      continue # to be removed before merge
```
this can be removed

cjac commented:
I met with customer to exercise the cluster and they asked for one additional change.  If the correct metadata key/value pair are passed, we now install pytorch and register a new kernel for use with JupyterLab.

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cnauroth commented on gpu/install_gpu_driver.sh:
```diff
@@ -1417,33 +2074,65 @@ function mount_ramdisk(){
+    grep -iv "^${feature} " /etc/ssh/sshd_config > /tmp/sshd_config_new
+    echo "$sshd_config_line" >> /tmp/sshd_config_new
+    # TODO: test whether sshd will reload with this change before mv
+    mv /tmp/sshd_config_new /etc/ssh/sshd_config
```
Update to `mv -f`.

cnauroth commented on gpu/install_gpu_driver.sh:
```diff
@@ -1453,15 +2142,145 @@ function prepare_to_install(){
+    if test -v DATAPROC_VERSION ; then
+      DATAPROC_IMAGE_VERSION="${DATAPROC_VERSION}"
+    else
+      if   version_lt "${SPARK_VERSION}" "3.2" ; then DATAPROC_IMAGE_VERSION="2.0"
```
Are there any cases where neither `DATAPROC_IMAGE_VERSION` nor `DATAPROC_VERSION` is set so we need to do this detection based on Spark version? I would not have expected this.

cjac replied:
Yes.  When running the init action scripts as custom-images input

cnauroth replied:
I hadn't thought of that use case. Can you please add a comment that the final else path is meant to support custom image builds?

cjac replied:
sure thing

cnauroth commented on gpu/install_gpu_driver.sh:
```diff
@@ -1453,15 +2142,145 @@ function prepare_to_install(){
+
+  curl -s -L "${repo_url}" \
+    | dd of="${repo_path}" status=progress
+#    | perl -p -e "s{^gpgkey=.*$}{gpgkey=file://${kr_path}}" \
```
Can this commented-out code be removed?

cjac replied:
https://issuetracker.google.com/issues/393219117

cnauroth reviewed (changes requested):
Thanks for this, @cjac ! I entered a few comments. In general, can we also review the `curl` calls to make sure they have the necessary retries?

cjac commented on gpu/test_gpu.py:
```diff
@@ -64,125 +152,147 @@ def verify_instance_spark(self):
                                      driver_provider):
-    if ( self.getImageOs() == 'rocky' ) and self.getImageVersion() >= pkg_resources.parse_version("2.2"):
-      self.skipTest("GPU drivers are currently FTBFS on Rocky 9 ; base dataproc image out of date")
+    self.skipTest("No need to regularly test installing the agent on its own cluster ; this is exercised elsewhere")
```
I could go either way.  It feels excessive to me.  This is an expensive test that doesn't really exercise another code path.  I test with install-gpu-agent=true in my reproduction environment during development.

What do you think?

cnauroth replied:
If it's exercised elsewhere, then should we remove the test entirely instead of skipping it?

cjac replied:
that seems reasonable

cjac commented on gpu/install_gpu_driver.sh:
```diff
@@ -970,13 +1446,12 @@ function configure_yarn_nodemanager() {
+  if version_ge "${SPARK_VERSION}" "3.0" ; then return 0 ; fi
+  # include exclusive mode on GPU
+  nvsmi -c EXCLUSIVE_PROCESS
+  clear_nvsmi_cache
```
good catch.  I think I accidentally merged this line in from my template work.  I've got some code to cache results of nvidia-smi to XML to avoid excessively requesting the same information.  Too many executions of nvidia-smi tends to bog the system down.  I've upated the getGpusResources.sh script to cache the results, too, since the value never changes ; no GPUs are attached or removed during node lifetime.

cjac commented on gpu/install_gpu_driver.sh:
```diff
@@ -1014,21 +1490,55 @@ function configure_gpu_script() {
+
+set -e
+resources_json="/dev/shm/nvidia/gpusResources.json"
+if test -f "${resources_json}" ; then cat "${resources_json}" ; exit 0 ; fi
```
this is to cache the result of nvidia-smi.  The result does not change, and each request is a heavy operation on the GPU.  Rather than running nvidia-smi every time that yarn wants to know about GPU resources, we instead cat the file with the results of the first request.

cnauroth replied:
Got it, I missed the broader context that this is ultimately executed inside the GPU discovery script, which we are generating here.

cjac reviewed (commented):
Thank you for the review, Chris and Prince.  I will make these changes and resolve each.

cnauroth commented on gpu/test_gpu.py:
```diff
@@ -6,25 +6,85 @@
-    time.sleep(3) # Many failed nvidia-smi attempts have been caused by impatience
+    import random
+    # Many failed nvidia-smi attempts have been caused by impatience and temporal collisions
+    time.sleep( 3 + random.randint(1, 30) )
```
Oh, now I get it! Thanks for clarifying.

cjac commented on gpu/install_gpu_driver.sh:
```diff
@@ -668,47 +922,36 @@ function add_nonfree_components() {
 
 function add_repo_cuda() {
   if is_debuntu ; then
-    local kr_path=/usr/share/keyrings/cuda-archive-keyring.gpg
```
I did patch it back in.

cjac commented:
/gcbrun

cjac commented:
@cnauroth - I believe that I addressed all of your concerns in the previous change, https://github.com/GoogleCloudDataproc/initialization-actions/pull/1275/commits/eacb99f5d57ea4a3a993a181747d87071ca73108

Please let me know if you see anything that needs further attention.

cnauroth commented on gpu/install_gpu_driver.sh:
```diff
@@ -134,13 +134,13 @@ readonly ROLE
 # Minimum supported version for open kernel driver is 515.43.04
 # https://github.com/NVIDIA/open-gpu-kernel-modules/tags
-latest="$(curl -s https://us.download.nvidia.com/XFree86/Linux-x86_64/latest.txt | awk '{print $1}')"
+#latest="$(curl -s https://us.download.nvidia.com/XFree86/Linux-x86_64/latest.txt | awk '{print $1}')"
```
Remove this line if `latest` is no longer being used?

cnauroth commented on gpu/install_gpu_driver.sh:
```diff
@@ -625,11 +627,22 @@ function install_nvidia_nccl() {
-      while gsutil ls "${gcs_tarball}.building" 2>&1 | grep -q "${gcs_tarball}.building" ; do
-        sleep 5m
-      done
+      if gsutil ls -j "${gcs_tarball}.building" > "${local_tarball}.building.json" ; then
```
Is `-j` a recent new feature? I couldn't find this option in my `gsutil`.

cjac replied:
it dumps the object metadata in JSON format ; I thought it would be `--format json` to match other gcloud commands, but I guess that it's slightly different because of gsutil.  The argument is documented for `gcloud storage ls --help` which I assumed used the same ABI as gsutil, but I don't see the argument in the `gcloud ls --help` documentation.

cjac replied:
as for whether it's new, I don't think it is.  I looked through the release history[1], and it seems that JSON support has been included from the beginning.  But I couldn't find the argument parsing code where the argument is detected on a quick review of [2].

[1] https://github.com/GoogleCloudPlatform/gsutil/blob/master/CHANGES.md
[2] https://github.com/GoogleCloudPlatform/gsutil

cjac commented:
/gcbrun

cjac commented:
/gcbrun

cjac commented:
2.2.44-rocky9 is able to build the kernel driver, so I've removed the test skip for rocky9.

cnauroth reviewed (approved):
@cjac , that addresses all of my feedback. Thank you very much!

LGTM, pending a successful CI run on the latest changes.

@prince-cs and @kuldeepkk-dev , please let us know if you had any other feedback.
